text/0005-media-webrtc-sdk.md

```diff
@@ -0,0 +1,708 @@
+| SDP_ERROR             | The sdp is invalid.     |
+| INVALID_REQUEST       | The request is invalid. |
+| INTERNAL_SERVER_ERROR | The server is error.    |
+| GATEWAY_ERROR         | The gateway is error.   |
+
+After that, the client needs to wait for the connected event from the WebRTC connection and the connected event from the data channel.
+If the client doesn't receive any event after a period of time, it will set the restart ice flag and retry connecting to the server with the newest offer SDP.
+After several attempts (configurable), the client will stop retrying and report an error to the user as CONNECTION_TIMEOUT.
+
+## 4.3 Restart-ice
+
+**_Endpoint_**: POST `GATEWAY/webrtc/:conn_id/restart-ice`
+
+**_Headers_**: `Authorization: Bear {token}`
+
+**_Body_**: same with connect request but the tracks should sending with current state
+
+**_Response Data_**: same with connect response
+
+By doing this, in case of a network change, the client can retry connecting to the server with the newest offer SDP. If the server is still alive, it will respond with a new answer SDP. However, if the server is dead, the gateway will retry connecting to another server. The session state can be restored using the track state and each feature state.
+
+## 4.4 Ice-tricle
+
+Each time the client's WebRTC connection has a new ice-candidate, it should be sent to the gateway using the following endpoint:
+
+**_Endpoint_**: POST `GATEWAY/webrtc/conns/:conn_id/ice-candidate`
+
+**_Body_**: candidate String
+
+**_Response Data_**: None
+
+## 4.5 Datachannel Request/Response format
+
+The format for encoding all requests and responses sent over the data channel is JSON. The structure of the request and response objects is as follows:
+
+Request/Event:
+
+```
+{
+    type: "event" | "request",
+    seq: Number,
+    cmd: String,
+    data: Option<JSON>,
+}
+```
+
+Response:
+
+```
+{
+    type: "answer",
+    seq: Number,
+    success: bool,
+    error_code: Option<String>,
+    error_msg: Option<String>,
+    data: Option<JSON>,
+}
+```
+
+The seq is an incremental value generated on the sender side. It helps us to map between requests and responses and also detect data loss.
+
+The cmd is generate with rule: `identify.action`, for example:
+
+- `peer.update_sdp`.
+- `sender.{sender_id}.toggle`.
+- `receiver.{receiver_id}.switch`.
+- `room.peers.subscribe`.
+- `room.peers.unsubscribe`.
+- `session.disconnect`.
+- `session.features.mix_minus.sources.add`.
+
+## 4.6 In-session requests
+
+At the current state, we already have one WebRTC connection to the server, so there is no need to send any requests over HTTP. All requests will be sent over the WebRTC datachannel.
+
+Typically, the client will need to perform various actions with the media server, such as:
+
+- Creating/Releasing senders
+- Creating/Releasing receivers
+- Sender actions: pause, resume, switch stream
+- Receiver actions: pause, resume, switch remote source, update priority, and layers
```

Comment:
these actions are not fully mentioned, for example, pause and resume action.

Reply:
pause and resume is simply `session.senders.{id}.toggle` with active flag

Reply:
You should mention this in the toggle actions.